import "ent/acoes.proto";
import "ent/comum.proto";
import "ent/entidade.proto";
import "ent/tabuleiro.proto";

package ntf;
option java_package = "com.matferib.Tabuleiro.ntf";

enum Tipo {
  TN_SAIR = 1;
  TN_ERRO = 2;
  TN_INFO = 30;
  TN_INICIAR = 3;  // Iniciar jogo modo mestre.
  TN_INICIADO = 4;
  TN_CONECTAR = 5;  // Conectar no jogo mestre.
  TN_CONECTAR_PROXY = 60;  // Mestre conectando no proxy.
  TN_DESCONECTAR = 6;  // Comando para desconectar jogo.
  TN_RESPOSTA_CONEXAO = 7;
  TN_DESCONECTADO = 8;  // Cliente ou servidor desconectou.
  TN_ADICIONAR_JOGADOR = 9;
  TN_ADICIONAR_ENTIDADE = 10;
  TN_LIMPAR_SALVACOES = 11;
  TN_REMOVER_ENTIDADE = 12;
  TN_ATUALIZAR_ENTIDADE = 13;
  TN_ATUALIZAR_PARCIAL_ENTIDADE = 29;
  TN_ATUALIZAR_TABULEIRO = 14;
  TN_ATUALIZAR_OPCOES = 23;
  TN_TEMPORIZADOR = 16;
  TN_SERIALIZAR_TABULEIRO = 17;
  TN_DESERIALIZAR_TABULEIRO = 18;
  TN_CRIAR_CENARIO = 41;
  TN_REMOVER_CENARIO = 42;
  TN_SERIALIZAR_ENTIDADES_SELECIONAVEIS = 32;
  TN_DESERIALIZAR_ENTIDADES_SELECIONAVEIS = 33;
  TN_MOVER_ENTIDADE = 19;
  TN_ABRIR_DIALOGO_ENTIDADE = 20;
  TN_ABRIR_DIALOGO_PROPRIEDADES_TABULEIRO = 21;
  TN_ABRIR_DIALOGO_OPCOES = 22;
  TN_ABRIR_DIALOGO_SALVAR_TABULEIRO_SE_NECESSARIO_OU_SALVAR_DIRETO = 44;
  TN_ABRIR_DIALOGO_SALVAR_TABULEIRO = 40;
  TN_ABRIR_DIALOGO_ABRIR_TABULEIRO = 45;
  TN_ABRIR_DIALOGO_MODELO_ENTIDADE = 56;
  TN_ABRIR_DIALOGO_COR_PERSONALIZADA = 57;
  TN_CARREGAR_TEXTURA = 24;
  TN_CARREGAR_MODELO_3D = 49;
  TN_DESCARREGAR_MODELO_3D = 51;
  TN_DESCARREGAR_TEXTURA = 25;
  TN_ADICIONAR_ACAO = 26;
  TN_REINICIAR_TABULEIRO = 27;
  TN_REINICIAR_CAMERA = 28;
  TN_SALVAR_CAMERA = 31;
  TN_ATUALIZAR_RODADAS = 35;  // Notifica clientes sobre atualizacao de rodadas.
  TN_ENVIAR_ID_TEXTURAS_E_MODELOS_3D = 37;   // Local: pedido para texturas enviar os ids locais para servidor.
  TN_REQUISITAR_MODELOS_3D = 36;  // Remota: cliente pedindo modelos 3d para o servidor, incluindo os que ja tem.
  TN_REQUISITAR_TEXTURAS = 38;  // Remota: cliente pedindo texturas para servidor, incluindo as que ja tem.
  TN_ENVIAR_TEXTURAS = 39;      // Servidor enviando texturas para cliente.
  TN_ENVIAR_MODELOS_3D = 50;      // Servidor enviando modelos 3d para cliente.
  TN_ALTERAR_MODO_MESTRE_SECUNDARIO = 43;  // Servidor pedindo para alterar o modo mestre secundario.
  TN_REFRESCAR_MENU = 46;  // Tabuleiro notificando que mudou algum modo e o menu deve se atualizar.
  TN_GERAR_TERRENO_ALEATORIO = 47;  // Notifica geracao de terreno aleatorio.
  TN_GERAR_MONTANHA = 55;  // Notifica geracao de terreno aleatorio.
  TN_ATUALIZAR_RELEVO_TABULEIRO = 48;        // indica atualizacao de terreno.
  TN_ATUALIZAR_LISTA_INICIATIVA = 52;  // Atualizacao da lista de iniciativas.
  TN_PROXIMA_INICIATIVA = 59;          // cliente requisitando passar a vez.
  TN_ENTRAR_MODO_SELECAO_TRANSICAO = 58;
  TN_ABRIR_DIALOGO_ESCOLHER_POCAO = 61;
  TN_HACK_ANDROID = 99;
  // O objetivo principal do grupo eh agrupar acoes locais. Nunca se deve enviar um grupo para a central.
  TN_GRUPO_NOTIFICACOES = 100;
}

// Por padrao, toda notificacao eh processada localmente e nao remotamente.
message Notificacao {
  optional Tipo tipo = 1;
  // Se verdadeiro, indica que deve ser enviada apenas para clientes pendentes. Toda notificacao deste tipo
  // devera ser acompanhada do id do cliente.
  // Caso contrario, enviado para todo pendentes.
  optional bool clientes_pendentes = 2 [default=false];
  // Se verdadeira, indica que o servidor nao deve repassar a mensagem para outros clientes.
  optional bool servidor_apenas = 17 [default=false];
  // Usado para varios tipos de respostas. Ausente quando nao ha erro.
  optional string erro = 4;
  // Identificacao propria do cliente no formato de string. Mensagens enviadas do servidor com este campo atribuido.
  // serao enviadas apenas para o cliente especificado.
  optional string id_rede = 15;
  // URL do servidor para conexao ou caminho de arquivo para serializacao deserializacao.
  // Muitas vezes usada para evitar repeticao de mensagens (cliente recebe e repassa ao servidor
  // que processa e repassa ao cliente infinitamente). Quando usada assim, a ausencia de endereco
  // significa, processe e nao repasse.
  optional string endereco = 5;
  // Tabuleiro de jogo.
  optional ent.TabuleiroProto tabuleiro = 6;
  // Para desfazer.
  optional ent.TabuleiroProto tabuleiro_antes = 16;
  // Entidade.
  optional ent.EntidadeProto entidade = 7;
  // Para atualizacoes de entidade, guardando o estado anterior a modificacao.
  optional ent.EntidadeProto entidade_antes = 3;
  // Usado na criacao de entidades.
  optional uint32 id_referencia = 19;
  // Acoes.
  optional ent.AcaoProto acao = 12;
  // Opcoes.
  optional ent.OpcoesProto opcoes = 8;
  // Informacao sobre textura para carregamento e descarregamento.
  repeated ent.InfoTextura info_textura = 9;
  // Informacao sobre modelos 3d.
  repeated ent.Modelo3d info_modelo_3d = 18;
  // Indica se o jogo esta em modo mestre ou nao.
  optional bool modo_mestre = 10;
  // Indica que a mensagem eh local, ou seja, veio do proprio jogo e nao de um cliente remoto.
  // Este campo eh preenchido para false durante recebimento de notificacoes pela rede.
  optional bool local = 11 [default=true];
  // Para notificacoes do tipo TN_GRUPO_NOTIFICACOES. Apenas alguns tipos de notificacoes sao suportadas.
  repeated Notificacao notificacao = 14;
}
